lib/ansible/galaxy/collection.py

```diff
@@ -0,0 +1,259 @@
+                           % (to_native(galaxy_yml_path), to_native(err)))
+
+    set_keys = set(galaxy_yml.keys())
+    missing_keys = mandatory_keys.difference(set_keys)
+    if len(missing_keys) > 0:
+        raise AnsibleError("The collection galaxy.yml at %s is missing the following mandatory keys: %s"
+                           % (to_native(galaxy_yml_path), ", ".join(missing_keys)))
+
+    extra_keys = set_keys.difference(all_keys)
+    if len(extra_keys) > 0:
+        display.warning("Found unknown keys in collection galaxy.yml at %s: %s"
+                        % (to_native(galaxy_yml_path), ", ".join(extra_keys)))
+
+    # Add the defaults if they have not been set
+    for optional_string in optional_strings:
+        if optional_string not in galaxy_yml:
+            galaxy_yml[optional_string] = ""
+
+    for optional_list in optional_lists:
+        list_val = galaxy_yml.get(optional_list, None)
+
+        if list_val is None:
+            galaxy_yml[optional_list] = []
+        elif not isinstance(list_val, list):
+            galaxy_yml[optional_list] = [list_val]
+
+    for optional_dict in optional_dicts:
+        if optional_dict not in galaxy_yml:
+            galaxy_yml[optional_dict] = {}
+
+    return galaxy_yml
+
+
+def _build_files_manifest(collection_path):
+    ignore_files = frozenset(['*.pyc', '*.retry'])
+    ignore_dirs = frozenset(['CVS', '.bzr', '.hg', '.git', '.svn', '__pycache__', '.tox'])
+
+    entry_template = {
+        'name': None,
+        'ftype': None,
+        'chksum_type': None,
+        'chksum_sha256': None,
+        '_format': MANIFEST_FORMAT
+    }
+    manifest = []
+
+    def _walk(path, top_level_dir, parent_dirs):
+        for item in os.listdir(path):
+            abs_path = os.path.join(path, item)
+            rel_base_dir = '' if path == top_level_dir else path[len(top_level_dir) + 1:]
+            rel_path = os.path.join(rel_base_dir, item)
+
+            if os.path.isdir(abs_path):
+                if item in ignore_dirs:
+                    display.vvv("Skipping %s for collection build" % abs_path)
+                    continue
+
+                if os.path.islink(abs_path):
+                    link_target = os.path.realpath(abs_path)
+
+                    if not link_target.startswith(top_level_dir):
+                        display.warning("Skipping %s as it is a symbolic link to a directory outside the collection"
+                                        % abs_path)
+                        continue
+
+                manifest_entry = entry_template.copy()
+                manifest_entry['name'] = rel_path
+                manifest_entry['ftype'] = 'dir'
+
+                manifest.append(manifest_entry)
+
+                _walk(abs_path, top_level_dir, parent_dirs)
+            else:
+                if item == 'galaxy.yml':
+                    continue
+                elif any([fnmatch.fnmatch(item, pattern) for pattern in ignore_files]):
```

Comment:
Casting to list isn't needed here
```suggestion
                elif any(fnmatch.fnmatch(item, pattern) for pattern in ignore_files):
```